File: backend/models/incident.py
"""
Pydantic models for security incidents and analysis
"""
from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any
from datetime import datetime
from enum import Enum


class SeverityLevel(str, Enum):
    """Incident severity levels"""
    LOW = "LOW"
    MEDIUM = "MEDIUM"
    HIGH = "HIGH"
    CRITICAL = "CRITICAL"


class IncidentStatus(str, Enum):
    """Incident lifecycle status"""
    DETECTED = "DETECTED"
    ANALYZING = "ANALYZING"
    ANALYZED = "ANALYZED"
    REMEDIATING = "REMEDIATING"
    REMEDIATED = "REMEDIATED"
    DOCUMENTED = "DOCUMENTED"
    CLOSED = "CLOSED"


class TimelineEvent(BaseModel):
    """A single event in the incident timeline"""
    timestamp: datetime
    actor: str
    action: str
    resource: str
    details: Optional[str] = None
    significance: Optional[str] = None
    severity: Optional[SeverityLevel] = None
    raw_event: Optional[Dict[str, Any]] = None


class Timeline(BaseModel):
    """Complete timeline analysis of an incident"""
    events: List[TimelineEvent]
    root_cause: Optional[str] = None
    attack_pattern: Optional[str] = None
    blast_radius: Optional[str] = None
    confidence: float = Field(ge=0.0, le=1.0)
    analysis_summary: Optional[str] = None


class CloudTrailEvent(BaseModel):
    """CloudTrail event structure"""
    eventTime: str
    eventName: str
    userIdentity: Dict[str, Any]
    requestParameters: Optional[Dict[str, Any]] = None
    responseElements: Optional[Dict[str, Any]] = None
    sourceIPAddress: Optional[str] = None
    resources: Optional[List[Dict[str, Any]]] = None
    errorCode: Optional[str] = None
    errorMessage: Optional[str] = None


class AnalysisRequest(BaseModel):
    """Request to analyze CloudTrail events"""
    events: List[Dict[str, Any]]
    incident_type: Optional[str] = None
    resource_id: Optional[str] = None


class AnalysisResponse(BaseModel):
    """Response from timeline analysis"""
    incident_id: str
    timeline: Timeline
    analysis_time_ms: int
    model_used: str


class Incident(BaseModel):
    """Complete incident record"""
    incident_id: str
    title: str
    description: str
    severity: SeverityLevel
    status: IncidentStatus
    detected_at: datetime
    timeline: Optional[Timeline] = None
    metadata: Optional[Dict[str, Any]] = None

File: frontend/src/types/incident.ts
/**
 * TypeScript types for incidents and timeline analysis
 */

export type SeverityLevel = 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';

export type IncidentStatus = 
  | 'DETECTED' 
  | 'ANALYZING' 
  | 'ANALYZED' 
  | 'REMEDIATING' 
  | 'REMEDIATED' 
  | 'DOCUMENTED' 
  | 'CLOSED';

export interface TimelineEvent {
  timestamp: string;
  actor: string;
  action: string;
  resource: string;
  details?: string;
  significance?: string;
  severity?: SeverityLevel;
  raw_event?: any;
}

export interface Timeline {
  events: TimelineEvent[];
  root_cause?: string;
  attack_pattern?: string;
  blast_radius?: string;
  confidence: number;
  analysis_summary?: string;
}

export interface AnalysisRequest {
  events: any[];
  incident_type?: string;
  resource_id?: string;
}

export interface AnalysisResponse {
  incident_id: string;
  timeline: Timeline;
  analysis_time_ms: number;
  model_used: string;
}

export type AgentStatus = 'PENDING' | 'RUNNING' | 'COMPLETED' | 'FAILED' | 'SKIPPED';

export interface AgentState {
  status: AgentStatus;
  started_at?: string;
  error?: string;
  reason?: string;
  model?: string;
}

export interface OrchestrationResponse {
  incident_id: string;
  status: 'completed' | 'failed' | 'running';
  analysis_time_ms: number;
  model_used?: string;
  agents: {
    temporal?: AgentState;
    visual?: AgentState;
    risk_scorer?: AgentState;
    remediation?: AgentState;
    documentation?: AgentState;
  };
  results: {
    timeline?: Timeline;
    visual?: any;
    risk_scores?: Array<{ event: string; risk?: any; risk_score?: number }>;
    remediation_plan?: any;
    documentation?: any;
    correlation?: {
      correlation_summary: string;
      campaign_probability: number;
      pattern_matches?: number;
      technique_overlaps?: number;
    };
  };
  metadata?: {
    incident_type?: string;
    [key: string]: any;
  };
}

export interface DemoScenario {
  id: string;
  name: string;
  description: string;
  severity: SeverityLevel;
  event_count: number;
}
